action: allow client-menu to open at mouse cursor

ahesford opened the pull request:
This is an alternative to #1604 that implements my desire to make this a per-action argument. Given that `at_cursor` is ignored for `root-menu`, this could also be a separate option, *e.g.* `ShowClientMenuHere`, which takes no arguments.

klausman commented:
FWIW, this approach is fine by me, I have no particular attachment to how #1604 works.

Consolatis commented on docs/labwc-actions.5.scd:
```diff
@@ -105,8 +105,15 @@ Actions are used in menus and keyboard/mouse bindings.
+*<action name="ShowMenu" menu="value" at_cursor="yes" />*
+	Show a menu.
+
+	*menu* The name of the menu to show, which must be one of "root-menu" or
+	"client-menu".
```
This is wrong, it can be named anything as long its define in `menu.xml`

Consolatis commented on src/action.c:
```diff
@@ -321,6 +321,10 @@ action_arg_from_xml_node(struct action *action, const char *nodename, const char
 			action_arg_add_str(action, argument, content);
 			goto cleanup;
 		}
+		if (!strcmp(argument, "at_cursor")) {
```
I think we usually use `CamelCase` for config args so `atCursor` might be more consistent.

johanmalm commented:
Nice work @ahesford 

Just thinking... would it make sense to let `atPointer` default to `yes` to clean up `show_menu()` a lot... making the `alt-space` keybind case the special one.

Looks like both `(!strcasecmp(menu_name, "client-menu")` and `ssd_part_contains(LAB_SSD_PART_TITLEBAR, type)` would go away with introducing this option.

We just have to define
- the cursor-button-press on Titlebar to have atPointer==true
- the alt-space key-bind to asPointer==true

ahesford commented:
In the first of two fixups, I:
1. Renamed `at_cursor -> atCursor` in the configuration
2. Adjusted the documentation to reflect arbitrary menu naming
3. Refactored `show_menu` a bit to take advantage of the new option

In the second fixup, I overhauled the definition of default key and mouse bindings to accommodate the desire to specify the `atCursor` attribute in a couple of `ShowMenu` actions. I considered doing something even more general, like parsing inline XML rather than modifying the combo structs, but that seemed like a much more substantive effort.

This change to the storage format makes declaration of the defaults a bit more cumbersome, but it should make it easy to add extra attributes should the defaults demand them in the future. I've looked at the C99 spec and believe that brace-initialization rules ensure that the attributes array in each struct is NULL terminated, so we should not explicitly require an all-NULL entry in cases where we define at least one attribute. The NULLs that do exist were required to get clang to be quiet about unspecified values. If this refactor is acceptable, we should do a bit more investigation to ensure that the fields really are NULL-terminated.

Consolatis commented on src/action.c:
```diff
@@ -321,6 +321,10 @@ action_arg_from_xml_node(struct action *action, const char *nodename, const char
 			action_arg_add_str(action, argument, content);
 			goto cleanup;
 		}
+		if (!strcmp(argument, "atCursor")) {
+			action_arg_add_bool(action, argument, parse_bool(content, true));
```
I think this has to be `strcasecmp()`, otherwise this will never match (as argument is lower-cased automatically afair)

ahesford replied:
You're right; this baffled me in testing for a few minutes. Fixed.

Consolatis commented on src/action.c:
```diff
@@ -571,42 +575,42 @@ action_list_free(struct wl_list *action_list)
 
-	int x, y;
-	if (force_menu_top_left) {
+	/* Place cursor in the view corner if desired (and menu is not root-menu) */
```
`s/Place cursor/Place  menu/` ?

ahesford replied:
Good catch.

Consolatis commented on src/config/rcxml.c:
```diff
@@ -1163,63 +1163,93 @@ rcxml_init(void)
-	{ "XF86_MonBrightnessUp", "Execute", "command", "brightnessctl set +10%" },
-	{ "XF86_MonBrightnessDown", "Execute", "command", "brightnessctl set 10%-" },
-	{ NULL, NULL, NULL, NULL },
+	{ "A-Tab", "NextWindow",
+		{ { NULL, NULL } } },
+	{ "W-Return", "Execute",
+		{ { "command", "alacritty" } } },
```
Personally I think this is slightly easier to read:
```c
	{ "A-Tab", "NextWindow", {
		{ NULL, NULL } }
	}, {  "W-Return", "Execute", {
		{ "command", "alacritty" } }
	}, {
		..
```

Also, my C foo is a bit lacking here, do we actually need the `{ NULL, NULL }` ones within the initializer? Usually everything not given should be set to a `0` or `NULL` equivalent but I am not sure if that is also happening for arrays within a struct.

ahesford replied:
I adopted your formatting. I tried various combinations of omitting empty values. GCC seems to require at least one field in the first struct of the array, so I was able to use `{ NULL }` instead of `{ NULL, NULL}`; clang complained about that and required at least one full struct initialization in the array before it was happy. The initialization as implemented seems to be the minimum necessary to keep the compilers happy.

Consolatis reviewed (commented):
Review below

Consolatis commented on src/action.c:
```diff
@@ -571,42 +575,34 @@ action_list_free(struct wl_list *action_list)
 
-	int x, y;
-	if (force_menu_top_left) {
+	/* Place menu in the view corner if desired (and menu is not root-menu) */
+	if (!at_cursor && strcasecmp(menu_name, "root-menu") != 0) {
 		x = view->current.x;
 		y = view->current.y;
```
This will break for custom menu names where there is no active view.
Do we actually need the exception for the "root-menu" here? If somebody sets it up to open as some kind of custom window menu, why not.

So I'd suggest something like this instead
```c
if (!at_cursor && view) {
	x = view->current.x;
	y = view->current.x;
}
```

ahesford replied:
Done

johanmalm commented on src/config/rcxml.c:
```diff
@@ -1163,63 +1163,94 @@ rcxml_init(void)
 {
 	struct keybind *k;
 	struct action *action;
+	struct key_combos *current;
```
Could reduce scope of `current` (put inside `(for (int i...) { ... }`

ahesford replied:
Moved in this function and in `load_default_mouse_bindings`.

johanmalm commented on src/config/rcxml.c:
```diff
@@ -1163,63 +1163,94 @@ rcxml_init(void)
-	{ "XF86_MonBrightnessUp", "Execute", "command", "brightnessctl set +10%" },
-	{ "XF86_MonBrightnessDown", "Execute", "command", "brightnessctl set 10%-" },
-	{ NULL, NULL, NULL, NULL },
+	{ "A-Tab", "NextWindow", {
```
Consider simplifying as follows:

```C
struct key_combos {
	const char *binding, *action;
	struct {
		const char *name, *value;
	} attributes[3];
} key_combos[] = { {
		.binding = "A-Tab",
		.action = "NextWindow",
	}, {
		.binding = "A-Space",
		.action = "ShowMenu",
		.attributes[0] = {
			.name = "menu",
			.value = "client-menu",
	       	},
		.attributes[1] = {
			.name = "atCursor",
			.value = "no",
	       	},
	}
};
```

ahesford replied:
This is, I think, much more readable than other options. I've adopted it int he latest commit.

johanmalm commented on src/config/rcxml.c:
```diff
@@ -1163,63 +1163,94 @@ rcxml_init(void)
+		{ "command", "brightnessctl set +10%" } }
+	}, { "XF86_MonBrightnessDown", "Execute", {
+		{ "command", "brightnessctl set 10%-" } }
+	}, { NULL, NULL, { { NULL, NULL } } },
```
The for-loop checks `.binding != NULL` so will be enough to set `.binding = NULL` for last entry.

ahesford replied:
Done

johanmalm reviewed (commented):
Looks good other than the comments.

I love the way `show_menu()` is so much simpler.

ahesford commented:
Although the latest form of defining `key_combos` and `mouse_combos` looks much cleaner and more obvious, it is also **long**. Thoughts on moving the static variables into `include/config/default-bindings.h` or something like that?

Consolatis commented:
> Although the latest form of defining `key_combos` and `mouse_combos` looks much cleaner and more obvious, it is also **long**. Thoughts on moving the static variables into `include/config/default-bindings.h` or something like that?

No objections from my side.

ahesford commented:
I moved the combo definitions and moved to using `ARRAY_SIZE` when iterating over attributes. This will help avoid bugs in the future if we decide to add a third attribtue to a combo and forget to bump the attribute array to 4 elements.

johanmalm commented:
Just looked through again. Really nice.
Thanks for working on this.
LGTM.
